### PR TITLE
[SPARK-17603] [SQL] Utilize Hive-generated Statistics For Partitioned Tables

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

For non-partitioned tables, Hive-generated statistics are stored in table properties. However, for partitioned tables, Hive-generated statistics are stored in partition properties. Thus, we are unable to utilize the Hive-generated statistics for partitioned tables. 

The statistics might not be gathered for all the partitions in Hive. For partial collection, we will not utilize the Hive-generated statistics.
### How was this patch tested?

Added test cases.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @hvanhovell @cloud-fan

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala [hunk 1]
```diff
@@ -109,39 +109,59 @@ private[hive] case class MetastoreRelation(
...
+            })
+        }
+      )
+    }
+  }
+
+  // For partitioned tables, get the size of all the partitions.
+  // Note: the statistics might not be gathered for all the partitions.
+  // For partial collection, we will not utilize the Hive-generated statistics.
+  private def getTotalTableSize(statType: String): Option[Long] = {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala]: what if a query only reads some partitions? Looks like the table statistics depend on partition pruning.

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala]: : ) We are facing the same issue in both data source tables and hive tables. That requires another PR to change how we use the statistics of leaf nodes. Partition filtering on statistics should be considered by the `Filter` nodes.  Is my understanding right?

**Comment by cloud-fan** [comment]
hmm, isn't is still valid?

**Pull request closed without merging**